api/envoy/api/v2/core/protocol.proto

```diff
@@ -46,6 +46,9 @@ message HttpProtocolOptions {
   // maximum number of request headers allowed is 100. Requests that exceed this limit will receive
   // a 431 response for HTTP/1.x and cause a stream reset for HTTP/2.
   google.protobuf.UInt32Value max_headers_count = 2 [(validate.rules).uint32 = {gte: 1}];
+
+  // Set transport socket SNI for new upstream connections based on HTTP host/authority header.
+  bool auto_sni = 4;
```

Comment:
Really sorry for just realizing this now, but I'm not thrilled that this upstream only option is being used on a common message which is shared both by downstream and upstream. A few options here:
1) Should this actually just be a config option on the router filter?
2) Should we have a separate message on the cluster specific to upstream only options?

@envoyproxy/api-shepherds thoughts here?

Reply:
@mattklein123
1. doesn't work for the use case, this really depends on whether upstream cluster support the SNI and/or the transport socket has SNI set. i.e. for a route mixing: a normal cluster with SNI override, and a original dst cluster.

2. That's reasonable, though we already have this pattern in `Http1/2ProtocolOptions` below, they are both mix of upstream and downstream protocol options. I think if things get confusing enough we should split all of the 3 `Http*ProtocolOptions` messages.

So we can do either 2 in this PR or just file an issue to split all 3 in v4alpha API.

Reply:
I would prefer to just do (2) now but if we need to get this landed, let's at least put a comment TODO in the API code to fix this?

/wait

Reply:
OK, does `UpstreamHttpProtocolOptions` sounds good to you? If so we can just add this new message contains only `auto_sni`. And the rest should be mechanical change.

Reply:
+1

Reply:
ok, I will fix with `UpstreamHttpProtocolOptions`